Changes to pdf page size handling

xiota opened the pull request:
Changes to `valid_size()`
* accept common page sizes, such as letter and a4.
* parse dimensions of format: AuxBv#, where A is width, u is units, B is height, v is units, # are options.
* accept units: in, cm, mm, pt

Changes to `convert()`:
* resize pages based on dpi calculations
* default resize images into page size (like default resize in imagemagick)
* implement exact resizing (ignore dpi; equiv to ! in imagemagick)

Created `get_ndpi()`:
* calculates dpi for page resizing
* implement fill resizing (equiv to ^ in imagemagick)

Other changes:
* default dpi in global variable

xiota commented:
@josch When you've had a chance to look this over, please let me know what you think.  Thanks.

josch commented on src/img2pdf.py:
```diff
@@ -347,21 +376,80 @@ def valid_date(string):
+    try:
+        string = papersizes[string.lower()]
+    except KeyError, e:
+        pass
```
Instead of this try/except you can just use:

    string = papersizes.get(string.lower(), string)

xiota replied:
will make appropriate changes

josch commented on src/img2pdf.py:
```diff
@@ -347,21 +376,80 @@ def valid_date(string):
+    except KeyError, e:
+        pass
+
+    pattern = re.compile('([0-9]*\.?[0-9]*)([a-z]*)x([0-9]*\.?[0-9]*)([a-z]*)([^0-9a-z]*)')
```
- If you are only using a pattern once, you do not need to compile it.
 - instead of `[0-9]` you can use `\d`
 - instead of `[a-z]` you might want to restrict the matches to all allowed units explicitly
 - instead of `[^0-9a-z]` you might want to list the allowed values instead

xiota replied:
* Will make changes to improve regex readability (multiline, as you suggest).
* Although the regex is still used only once, it will need to be compiled because changing to multiline involved setting the `re.VERBOSE` flag.
* I think `[0-9]` is more readable than `\d`, and the length of the regex is not and issue after making multiline.  Is there a performance benefit?
* Did not set units explicitly because it complicates the regex and would require changing the regex each time a new unit is added.  Code that interprets the units (not yet written) can display a warning.
* Similarly for the allowed extra options, do not want to have to change the regex every time a new option is added.  Code that interprets the options just searches for the correct symbol and sets a flag.  Is it a problem if extra junk is just ignored?

josch commented on src/img2pdf.py:
```diff
@@ -347,21 +376,80 @@ def valid_date(string):
+    # tokens.group(5) == extra options
+
+    if tokens.group(0) == '':
         msg = "input size needs to be of the format Ax, xB or AxB with A and B being integers"
```
This error message should explain the full format

xiota replied:
will make appropriate changes

josch commented on src/img2pdf.py:
```diff
@@ -347,21 +376,80 @@ def valid_date(string):
+    try:
+        x = float(tokens.group(1))
+    except ValueError, e:
         x = None
```
If something that matches the regex `[0-9]*\.?[0-9]*` cannot be parsed as float, then something went seriously wrong and you should throw an error accordingly.

xiota replied:
The regex also matches "." and the empty string.  In both cases, we really do want x or y to be assigned to None.

josch commented on src/img2pdf.py:
```diff
@@ -212,11 +213,26 @@ def tostring(self):
+
+    if pagesize[2]:
+        for key, value in pagesize_options.items():
+            pattern = re.compile(value[0])
```
the compilation of the regex should happen above so that there are not multiple compilations

xiota replied:
The regular expression is different depending on which option we are looking at (eg, fill), so it does need to be inside the loop.  Since the expressions are not reused, will change to use `re.search(pattern, string)`.

josch commented on src/img2pdf.py:
```diff
@@ -347,21 +376,80 @@ def valid_date(string):
-        y = int(y)
-    return (x,y)
+
+    return (x, y, tokens.group(5))
```
You should check the pagesize options for validity before returning them

xiota replied:
Will add checking to ensure that at least one x and y is assigned a value

josch commented on src/img2pdf.py:
```diff
@@ -320,21 +334,36 @@ def convert(images, dpi=None, pagesize=(None, None), title=None, author=None,
+            if pagesize[0]:
+                tmp_dpi = 72.0*width/pagesize[0]
+            else:
+                tmp_dpi = 72.0*width/pagesize[1]
```
What if pagesize[1] is None?

xiota replied:
At least one pagesize[0] or pagesize[1] should be assigned.  Will add checking in `valid_size()` to ensure that this is the case.

xiota commented:
@josch Have you had a chance to look at the changes since your previous comments?

Since you've moved to a different project host, should I make an account and submit a merge request there?

josch commented:
Hi,

Quoting xiota (2015-03-17 04:04:18)
> @josch Have you had a chance to look at the changes since your previous
> comments?
> 
> Since you've moved to a different project host, should I make an account and
> submit a merge request there?

no, that's not necessary. I'll review the changes here.

Also, you don't need to make a new account over there but can log in with your
github credentials.

I'll still take some time before I can take a look at your work because I'm
swamped with stuff to do over here.

cheers, josch

xiota commented:
@josch Figured out how to use git rebase, so the back and forth changes between commits are gone.  Removed default to pt when unit is unrecognized.  Shortened some long lines, but not sure which lines you commented about b/c can't find the comment after git rebase.

josch commented:
This looks much better, thanks!

Would it be possible for you to also write some examples about this syntax and what it use cases each of the modifiers have in the README.md?

xiota commented:
@josch Made three commits with changes to README.md:
 * "general editing" with rephrasing and rewriting throughout the entire file.
 * renamed "Commandline Options" to "Usage" with output from `img2pdf --help`
 * created new "Examples" section with subsections "controlling page size" and "colorspace"

Thinking... "Usage" should probably not be just the help output.  But the help output is pretty self explanatory for almost all of the options.  Will revisit it later (now that I know how to rebase).

josch commented:
Hi, this looks really great! Would you like me to already merge this pull request or would you rather like to add more commits first?

Maybe just remove the sections "Usage" or "Commandline options" altogether. I do not think that they are needed at all in the README.md as the `--help` output already gives this info (as you pointed out already). It would not be good to duplicate the info from `--help` in the README.md because it would easily lead to one of them being outdated.

May I ask whether you are a native english speaker?

xiota commented:
@josch English is my native language.  Most of the changes I made were to voice, so you could decide to keep it the way it is.  If English is not your native language, I don't think anyone would tell based on the comments I've seen you write.

Wouldn't have minded if you'd merged a long time ago, but it's much better after revising based on your comments and code reviews.  Probably will have similar positive results with README.md.

Deleted "Commandline Options" and changed "Examples" to "Usage and Examples".  The current subsections are:

 * general notes - explains why input cannot be piped through stdin and that output will be to stdout if not specified.
 * controlling page size
 * Colorspace - trying to explain the PIL short handles.  Found a reference that says ... `L` = grayscale, `1` = black and white, `RGB` and `CMYK` are what they sound like.  But what is `CMYK;I`?  `I` apparently refers to 32-bit "integer", but not sure how that ties in with CMYK.  Can they be reversed, like `I;CMYK` ?

Think after filling in the meaning of `CMYK;I`, it'll probably be ready to be merged.

josch commented:
Hi, english is not my native language. I only asked because since it is not my native language I am unable to tell whether your changes do sound better to the native ear or not. Since you are a native speaker, I'm happy to keep your changes to my sentences. Very much appreciated! Thanks!

What I see in your commits looks very good. Just tell me and I'll merge it with my codebase :)

xiota commented:
@josch I think it's good to merge.

josch commented:
Great! Committed here: https://gitlab.mister-muffin.de/josch/img2pdf/commits/master

zvezdochiot commented:
Hi all.

Maybe add options `DPI`?

```shell
-s "(width*72/DPI)x(height*72/DPI)"
```

Good Luck.

xiota commented:
This PR was closed a long time ago.  But if you run `img2pdf --help` with a current version, you'll see that it is already capable of using dpi:
```
  -s LxL, --imgsize LxL
                        Sets the size of the images on the PDF pages. In addition, the unit dpi
                        is allowed which will set the image size as a value of dots per inch.
                        Instead of a unit, width and height values may also have a percentage
                        sign appended, indicating a resize of the image by that percentage. The
                        short-option is the lower case s because it is an mnemonic for being
                        smaller than the page size.
```

zvezdochiot commented:
@xiota say:
>In addition, the unit dpi is allowed which will set the image size

Example?

xiota commented:
`img2pdf -s 600dpi -o example.pdf example.jpg`

zvezdochiot commented:
@xiota say:
>`img2pdf -s 600dpi -o example.pdf example.jpg`

Good. It's work. :+1:
